# Import Libraries

In [37]:
import pandas as pd
import numpy as np
import os

In [38]:
# Creating a dataframe
df_test = pd.DataFrame()

In [39]:
# Creating a mix-type column
df_test['mix'] = ['a', 'b', 1, True]

In [40]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [41]:
# Checking for mixed-type column
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [42]:
# Changing data type
df_test['mix'] = df_test['mix'].astype('str')

In [43]:
df_test.dtypes

mix    object
dtype: object

# Importing Data

In [44]:
# Importing data
path = '/Users/tannu/Desktop/Data Analytics/Instacart Basket Analysis Dec-2025 '
df_ords = pd.read_csv(os.path.join(path, '2 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '2 Data', 'Original Data', 'products.csv'), index_col = False)

In [45]:
# Checking missing values in products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [46]:
# Creating a subset to only view observations with missing values in product_name 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [47]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [48]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [49]:
df_prods['prices'].median()

7.1

In [50]:
df_prods.shape

(49693, 5)

In [51]:
# Creating new products dataframe with by excluding the observations with missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [52]:
df_prods_clean.shape

(49677, 5)

In [53]:
# Checking duplicates
df_prods_dups = df_prods_clean[df_prods_clean.duplicated()]

In [54]:
df_prods_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [55]:
# Creating a new dataframe which doesn't contain any duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [56]:
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting 'df_prods_clean_no_dups' 

In [57]:
# Exporting the clean products datframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '2 Data', 'Prepared Data', 'products_checked.csv'), index = False)

# Exercise 4.5

## Q: Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [58]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_the_week  order_hour_of_day  days_since_prior_order  
count            3.421083e+06       3.421083e+06            3.214874e+06  
mean             2.776219e+00       1.345202e+01            1.111484e+01  
std              2.046829e+00       4.226088e+00            9.206737e+00  
min              0.000000e+00       0.000000e+00            0.000000e+00  
25%              1.000000e+00       1.000000e+01            4.000000e+00  
50%              3.000000e+00       1.300000e+01            7.000000e+00  
75%              5.000000e+00       1.600000e+01            1.500000e+01  
max              6.000000e+00       2.300000e+01            3.000000e+01

### A: The data looks fine to me and I do not think there is any outlier value in the columns with numerical value.

## Q: Check for mixed-type data in your df_ords dataframe.

In [59]:
# Checking for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### A: There are no columns with mixed-type data in the df_ords dataframe.

## Q: Run a check for missing values in your df_ords dataframe.

In [60]:
# Checking for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### A: There are 206209 missing values in the column 'days_since_prior_order'. But the reason for that could be that a new customer will have a blank entry for this column for their first order. We can confirm our theory by checking the total number of customers. The variable for that is 'user_id'.

## Q: Address the missing values using an appropriate method.

In [61]:
df_ords['user_id'].describe()

count    3.421083e+06
mean     1.029782e+05
std      5.953372e+04
min      1.000000e+00
25%      5.139400e+04
50%      1.026890e+05
75%      1.543850e+05
max      2.062090e+05
Name: user_id, dtype: float64

### A: Since the maximum number of customers is 206209, the same as the missing values in 'days_since_prior_order', we can leave the missing values as it is.

## Q: Run a check for duplicate values in your df_ords data.

In [62]:
# Checking for duplicates
df_dups = df_ords[df_ords.duplicated()]

### A: There are no duplicates in this dataframe.

#### Since there is no change in the 'orders_wrangled.csv' file, I am not exporting it. I have already exported clean file for 'products' dataframe above.

In [63]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_the_week  \
0           0   2539329        1             1                       2   
1           1   2398795        1             2                       3   
2           2    473747        1             3                       3   
3           3   2254736        1             4                       4   
4           4    431534        1             5                       4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [64]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [65]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [66]:
df_ords.to_csv(os.path.join(path, '2 Data', 'Prepared Data', 'orders_checked.csv'), index = False)

In [69]:
# importing 'orders_checked' dataframe
df_checked = pd.read_csv(os.path.join(path, '2 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

In [68]:
df_checked.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2398795        1             2                       3                  7   
2    473747        1             3                       3                 12   
3   2254736        1             4                       4                  7   
4    431534        1             5                       4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0